include/circt/Query/Query.h

```diff
@@ -0,0 +1,217 @@
+#ifndef CIRCT_QUERY_API_H
+#define CIRCT_QUERY_API_H
+
+#include <iostream>
+#include <string>
+#include <unordered_map>
+#include <vector>
+
+#include "circt/Dialect/Comb/CombDialect.h"
+#include "circt/Dialect/Comb/CombVisitors.h"
+#include "circt/Dialect/HW/HWOps.h"
+#include "circt/Dialect/HW/HWTypes.h"
+#include "circt/Dialect/HW/HWVisitors.h"
+#include "circt/Dialect/SV/SVOps.h"
+#include "circt/Dialect/SV/SVVisitors.h"
+#include "llvm/Support/Regex.h"
+
+namespace circt {
+namespace query {
+
+class FilterType {
+public:
+  virtual ~FilterType() {}
+
+  virtual bool valueMatches(llvm::StringRef value) = 0;
+  virtual bool addSelf() { return false; }
```

Comment:
alternately, make this non-virtual, have it return a data field, then have subclasses set the data field at construction.